gax-httpjson/src/main/java/com/google/api/gax/httpjson/HttpJsonDirectStreamController.java

```diff
@@ -0,0 +1,126 @@
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.httpjson;
+
+import com.google.api.gax.rpc.ResponseObserver;
+import com.google.api.gax.rpc.StreamController;
+import com.google.common.base.Preconditions;
+import java.util.concurrent.CancellationException;
+
+/**
+ * Wraps a HttpJsonClientCall in a {@link StreamController}. It feeds events to a {@link
+ * ResponseObserver} and allows for back pressure.
+ *
+ * <p>Package-private for internal use.
+ */
+class HttpJsonDirectStreamController<RequestT, ResponseT> implements StreamController {
+  private final HttpJsonClientCall<RequestT, ResponseT> clientCall;
+  private final ResponseObserver<ResponseT> responseObserver;
+  private volatile boolean hasStarted;
+  private volatile boolean autoflowControl = true;
+  private volatile int numRequested;
+  private volatile CancellationException cancellationException;
+
+  HttpJsonDirectStreamController(
+      HttpJsonClientCall<RequestT, ResponseT> clientCall,
+      ResponseObserver<ResponseT> responseObserver) {
+    this.clientCall = clientCall;
+    this.responseObserver = responseObserver;
+  }
+
+  @Override
+  public void cancel() {
+    cancellationException = new CancellationException("User cancelled stream");
+    clientCall.cancel(null, cancellationException);
+  }
+
+  @Override
+  public void disableAutoInboundFlowControl() {
+    Preconditions.checkState(
+        !hasStarted, "Can't disable automatic flow control after the stream has started.");
+    autoflowControl = false;
+  }
+
+  @Override
+  public void request(int count) {
+    Preconditions.checkState(!autoflowControl, "Autoflow control is enabled.");
+
+    // Buffer the requested count in case the consumer requested responses in the onStart()
+    if (!hasStarted) {
+      numRequested += count;
+    } else {
+      clientCall.request(count);
+    }
```

Comment:
Can `request()` and `start()` ever run concurrently? If so, this isn't going to work as inteded unless there is a proper synchronization. (For example, on line 76, `hasStarted` may be false, but the moment you reach line 77, the value can be flipped.) Just to confirm this doesn't happen.

Reply:
This code, as most parts of this PR  mimicks grpc behavior, and this specific logic is from https://github.com/googleapis/gax-java/blob/main/gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcDirectStreamController.java#L89, which is not sycnhronized either. The grpc logic has been there for a few years already and no any race conditions problems so far. Basicaly this logic is externally synchronized. I made those variables volatile just in case to make sure thread visibility. Probably I need to revert them back to non-volatile ones to eliminate confusin.

Reply:
Yeah, maybe revert them back. I don't see them in the gRPC side either, and actually, it was those `volatile` keywords that made me to think about concurrency concerns in this class. Since you said this is externally synchronized, it should rather be explicit, I think.